internal/statemachine/helper.go

```diff
@@ -354,3 +371,150 @@ func maxOffset(offset1, offset2 quantity.Offset) quantity.Offset {
+		}
+
+		var structureType string
+		// Check for hybrid MBR/GPT
+		if strings.Contains(structure.Type, ",") {
+			types := strings.Split(structure.Type, ",")
+			if volume.Schema == "gpt" {
+				structureType = types[1]
+			} else {
+				structureType = types[0]
+			}
+		} else {
+			structureType = structure.Type
+		}
+
+		if volume.Schema == "mbr" {
+			bootable := false
+			if structure.Role == gadget.SystemBoot || structure.Label == gadget.SystemBoot {
+				bootable = true
+			}
+			// mbr.Type is a byte. snapd has already verified that this string
+			// is exactly two chars, so we can parse those two chars to a byte
+			partitionType, _ := strconv.ParseUint(structureType, 16, 8)
+			mbrPartition := &mbr.Partition{
+				Start:    uint32(*structure.Offset) / uint32(sectorSize),
+				Size:     uint32(structure.Size),
+				Type:     mbr.Type(partitionType),
+				Bootable: bootable,
+			}
+			mbrPartitions = append(mbrPartitions, mbrPartition)
+		} else {
+			partitionType := gpt.Type(structureType)
+			gptPartition := &gpt.Partition{
+				Start: uint64(*structure.Offset) / sectorSize,
+				Size:  uint64(structure.Size),
+				Type:  partitionType,
+				Name:  structure.Name,
+			}
+			gptPartitions = append(gptPartitions, gptPartition)
+		}
+	}
+
+	if volume.Schema == "mbr" {
+		mbrTable := &mbr.Table{
+			Partitions:         mbrPartitions,
+			LogicalSectorSize:  int(sectorSize),
+			PhysicalSectorSize: int(sectorSize),
+		}
+		partitionTable = mbrTable
+	} else {
+		gptTable := &gpt.Table{
+			Partitions:         gptPartitions,
+			LogicalSectorSize:  int(sectorSize),
+			PhysicalSectorSize: int(sectorSize),
+			ProtectiveMBR:      true,
+		}
+		partitionTable = gptTable
+	}
+	return &partitionTable
+}
+
+// calculateImageSize calculates the total sum of all partition sizes in an image
+func (stateMachine *StateMachine) calculateImageSize() (int64, error) {
+	if stateMachine.GadgetInfo == nil {
+		return 0, fmt.Errorf("Cannot calculate image size before initializing GadgetInfo")
+	}
+	var imgSize int64 = 0
+	for _, volume := range stateMachine.GadgetInfo.Volumes {
+		for _, structure := range volume.Structure {
+			imgSize += int64(structure.Size)
+		}
+	}
+	return imgSize, nil
+}
+
+// copyDataToImage runs dd commands to copy the raw data to the final image with appropriate offsets
+func (stateMachine *StateMachine) copyDataToImage(volumeName string, volume *gadget.Volume, diskImg *disk.Disk) error {
+	for structureNumber, structure := range volume.Structure {
+		if shouldSkipStructure(structure, stateMachine.isSeeded) {
+			continue
+		}
+		sectorSize := diskImg.LogicalBlocksize
+		// set up the arguments to dd the structures into an image
+		partImg := filepath.Join(stateMachine.tempDirs.volumes, volumeName,
+			"part"+strconv.Itoa(structureNumber)+".img")
+		seek := strconv.FormatInt(int64(getStructureOffset(structure))/sectorSize, 10)
+		count := strconv.FormatFloat(math.Ceil(float64(structure.Size)/float64(sectorSize)), 'f', 0, 64)
```

Comment:
...especially that here we seem to be doing the `math.Ceil()` dance. So let's get the partition part consistent too, I suppose?